Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp

```diff
@@ -263,4 +263,50 @@ void RenderTreeUpdater::GeneratedContent::removeAfterPseudoElement(Element& elem
     element.clearAfterPseudoElement();
 }
 
+void RenderTreeUpdater::GeneratedContent::updateWritingSuggestionsRenderer(RenderElement& renderer)
+{
+    auto destroyWritingSuggestionsIfNeeded = [&renderer, this]() {
+        if (WeakPtr writingSuggestionsRenderer = renderer.writingSuggestionsRenderer())
+            m_updater.m_builder.destroy(*writingSuggestionsRenderer);
+    };
+    if (!renderer.element()) {
+        destroyWritingSuggestionsIfNeeded();
+        return;
+    }
+
+    if (renderer.element()->writingSuggestion().isEmpty()) {
+        destroyWritingSuggestionsIfNeeded();
+        return;
+    }
+
+    if (!renderer.canHaveChildren())
+        return;
+
+    auto style = renderer.getCachedPseudoStyle({ PseudoId::InternalWritingSuggestions }, &renderer.style());
+    if (!style || style->display() == DisplayType::None) {
+        destroyWritingSuggestionsIfNeeded();
+        return;
+    }
+
+    if (renderer.writingSuggestionsRenderer() && renderer.writingSuggestionsRenderer()->nextSibling())
+        destroyWritingSuggestionsIfNeeded();
+
+    auto newStyle = RenderStyle::clone(*style);
+    if (auto writingSuggestionsRenderer = renderer.writingSuggestionsRenderer()) {
+        writingSuggestionsRenderer->setStyle(WTFMove(newStyle));
+
+        if (auto child = writingSuggestionsRenderer->firstChild())
+            uncheckedDowncast<RenderText>(child)->setText(renderer.element()->writingSuggestion());
+    } else {
+        auto newWritingSuggestionsRendererParent = WebCore::createRenderer<RenderInline>(RenderObject::Type::Inline, renderer.document(), WTFMove(newStyle));
+        newWritingSuggestionsRendererParent->initializeStyle();
+
+        auto newWritingSuggestionsRenderer = WebCore::createRenderer<RenderText>(RenderObject::Type::Text, renderer.document(), renderer.element()->writingSuggestion());
+        m_updater.m_builder.attach(*newWritingSuggestionsRendererParent, WTFMove(newWritingSuggestionsRenderer));
+
+        renderer.setWritingSuggestionsRenderer(*newWritingSuggestionsRendererParent.get());
+        m_updater.m_builder.attach(renderer, WTFMove(newWritingSuggestionsRendererParent));
```

Comment:
do you know if it works inside inline boxes with isolates and/or bidi in general?

Reply:
I do not know; do you know how I could test?

I know for some cases this does not work (ie, the suggestion doesn’t show up), however this approach is currently “best effort”, since suggestions may not appear anyways if there aren’t any, and this can be improved over time without having to disable the feature.

Reply:
also (though I think I asked about already) suggestion right after initial letter (not even sure if it's a thing) or just with general mutation e.g. programmatic content removal while editing.

Reply:
> suggestion right after initial letter
I haven’t been able to trigger a suggestion after a single letter 

> programmatic content removal while editing
do you have an example of this?

Reply:
> I do not know; do you know how I could test?

the first step would be just put contenteditable on a dir=rtl block

Reply:
yup, it works with rtl

Reply:
So the suggestion is always added at the end of the parent. What happens if we are editing in the middle of a text?

Reply:
Fixed!